engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -647,6 +680,138 @@ static void CheckTextureResource(lua_State* L, int i, const char* field_name, dm
+    dmGraphics::TextureImage::Type tex_type            = GraphicsTextureTypeToImageType(type);
+    dmGraphics::TextureImage::TextureFormat tex_format = GraphicsTextureFormatToImageFormat(format);
+
+    if (!dmGraphics::IsTextureFormatSupported(g_ResourceModule.m_GraphicsContext, format))
+    {
+        return luaL_error(L, "Unable to set texture, unsupported texture format '%s'.", dmGraphics::GetTextureFormatLiteral(format));
+    }
+
+    // TODO: To support this, we need to supply separate buffers for each side as an option, or offsets into the buffer where each side is located
+    if ((tex_type == dmGraphics::TextureImage::TYPE_CUBEMAP || tex_type == dmGraphics::TextureImage::TYPE_2D_ARRAY) && compression_type != dmGraphics::TextureImage::COMPRESSION_TYPE_DEFAULT)
+    {
+        return luaL_error(L, "Compression type %d requested for texture %s with type '%s', but this is currently not supported.",
+            (int) compression_type, path, dmGraphics::GetTextureTypeLiteral(type));
+    }
+
+    dmGameObject::HInstance sender_instance = dmScript::CheckGOInstance(L);
+
+    params->m_Path            = path;
+    params->m_PathHash        = path_hash;
+    params->m_Width           = width;
+    params->m_Height          = height;
+    params->m_MaxMipMaps      = max_mipmaps;
+    params->m_Type            = type;
+    params->m_Format          = format;
+    params->m_TextureBpp      = tex_bpp;
+    params->m_TextureType     = tex_type;
+    params->m_TextureFormat   = tex_format;
+    params->m_CompressionType = compression_type;
+    params->m_Buffer          = buffer;
+    params->m_Collection      = dmGameObject::GetCollection(sender_instance);
+    return 0;
+}
+
+static void HandleRequestCompleted(SetTextureAsyncRequest* request)
+{
+    // Swap out the texture
+    dmGraphics::DeleteTexture(request->m_TextureResource->m_Texture);
+    request->m_TextureResource->m_Texture = request->m_Texture;
+
+    if (dmScript::IsCallbackValid(request->m_CallbackInfo))
+    {
+        lua_State* L = dmScript::GetCallbackLuaContext(request->m_CallbackInfo);
+        DM_LUA_STACK_CHECK(L, 0);
+
+        // callback has the format:
+        // function(self, request_id, result)
+        if (dmScript::SetupCallback(request->m_CallbackInfo))
+        {
+            lua_pushnumber(L, request->m_Handle);
+
+            lua_newtable(L);
+            dmScript::PushHash(L, request->m_PathHash);
+            lua_setfield(L, -2, "path");
+
+            dmScript::PCall(L, 3, 0);
+            dmScript::TeardownCallback(request->m_CallbackInfo);
+        }
+        else
+        {
+            dmLogError("Failed to setup resource.create_texture_async callback (has the calling script been destroyed?)");
+        }
+
+        dmScript::DestroyCallback(request->m_CallbackInfo);
+    }
+
+    // If we used a temporary upload buffer, we need to delete it.
+    if (request->m_UseUploadBuffer)
+    {
+        dmBuffer::Destroy(request->m_Buffer);
+    }
```

Comment:
This is so we can support passing in a nil buffer to create a blank texture async